### PR TITLE
Add default credentials to images created by the ImageExtension

### PR DESCRIPTION
**Pull request by nacx**
By default, when listing images the ComputeServiceAdapter adds the default credentials for each image. This is not done when images are created by the image extension, and NPEs can appear in code that assumes the default credentials are there, as the field is not nullable.

This change tries to populate the known node credentials for images created form nodes, and falls back to the default strategy to add the default credentials to an image if there are no known credentials.

(Note this will require a similar PR in jclouds-labs to update other ComputeService constructors).

@danielestevez Can you try this commit and see if it fixes the issues you found with the ImageExtension?

### CODE REVIEW

File: compute/src/main/java/org/jclouds/compute/extensions/internal/DelegatingImageExtension.java [hunk 1]
```diff
@@ -49,6 +76,32 @@ public ImageTemplate buildImageTemplateFromNode(String name, String id) {
...
+               if (input.getDefaultCredentials() != null) {
+                  return input;
+               }
+               Credentials nodeCredentials = credentialStore.get("node#" + cloneImageTemplate.getSourceNodeId());
+               if (nodeCredentials != null) {
+                  logger.info(">> Adding node(%s) credentials to image(%s)...", cloneImageTemplate.getSourceNodeId(),
+                        cloneImageTemplate.getName());
+                  return ImageBuilder.fromImage(input)
+                        .defaultCredentials(LoginCredentials.fromCredentials(nodeCredentials)).build();
+               } else {
```
[the remark this answers is not in the document]
Comment (nacx) [reply compute/src/main/java/org/jclouds/compute/extensions/internal/DelegatingImageExtension.java]: Addressed in the last commit. Thanks for having a look!

**Comment by nacx** [comment]
>However testImageIsRemovedFromCacheAfterDeletion still fails it testSpawnNodeFromImage is enabled but that's a different issue.

I've just run the `AzureComputeImageExtensionLiveTest` in `northeurope` and with this change all pass:

```bash
Running org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest
Configuring TestNG with: TestNG652Configurator
Starting test testCreateImage(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest)
[pool-1-thread-1] Test testCreateImage(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest) succeeded: 423843ms
Test suite progress: tests succeeded: 1, failed: 0, skipped: 0.
Starting test testImageIsCachedAfterBeingCreated(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest)
Starting test testSpawnNodeFromImage(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest)
[pool-1-thread-2] Test testImageIsCachedAfterBeingCreated(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest) succeeded: 95495ms
Test suite progress: tests succeeded: 2, failed: 0, skipped: 0.
[pool-1-thread-3] Test testSpawnNodeFromImage(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest) succeeded: 442300ms
Test suite progress: tests succeeded: 3, failed: 0, skipped: 0.
Starting test testDeleteImage(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest)
[pool-1-thread-4] Test testDeleteImage(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest) succeeded: 10365ms
Test suite progress: tests succeeded: 4, failed: 0, skipped: 0.
Starting test testImageIsRemovedFromCacheAfterDeletion(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest)
[pool-1-thread-5] Test testImageIsRemovedFromCacheAfterDeletion(org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest) succeeded: 99238ms
Test suite progress: tests succeeded: 5, failed: 0, skipped: 0.
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1,183.506 sec - in org.jclouds.azurecompute.arm.compute.extensions.AzureComputeImageExtensionLiveTest

Results :

Tests run: 5, Failures: 0, Errors: 0, Skipped: 0
```

**Comment by nacx** [comment]
Well, yes and no :) @danielestevez is working on fixing jclouds to properly support Azure CanadaEast region. The issue is that the Azure API sometimes returns the region name lowercase, sometimes camelcase, etc, and that messes things up when it comes to filter by location and do several location comparisons jclouds does internally. We still don't know the real scope of this change, but I'm confident it is under control and it won't be a huge deal :)

The image without credentials thing is just isolated to this PR, so I think it is safe to merge (but I'll file a JIRA too).

**Comment by danielestevez** [comment]
Some context about what @nacx is talking about 
https://stackoverflow.com/questions/50068324/azure-arm-api-returns-locations-with-inconsistent-case

It affects also other of the new regions (not only CanadaEast) and working on it we discovered some unrelated issues like this one with the default credentials. As a separate JIRA from the case insensitive support for this provider.

**Comment by nacx** [comment]
FTR: [JCLOUDS-1421](https://issues.apache.org/jira/browse/JCLOUDS-1421)

**Comment by nacx** [comment]
Merged t [master](http://git-wip-us.apache.org/repos/asf/jclouds/commit/b76a594e) ([labs](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/331fd685)) and [2.1.x](http://git-wip-us.apache.org/repos/asf/jclouds/commit/3878e6a6) ([labs](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/fbf1f4bd)).

**Pull request closed without merging**